nuxeo-features/nuxeo-platform-audit/nuxeo-platform-audit-core/src/test/java/org/nuxeo/ecm/platform/audit/AbstractAuditStorageTest.java

```diff
@@ -0,0 +1,171 @@
+ *
+ * Contributors:
+ *     Gethin James
+ */
+package org.nuxeo.ecm.platform.audit;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.nuxeo.ecm.platform.audit.api.BuiltinLogEntryData.LOG_DOC_PATH;
+import static org.nuxeo.ecm.platform.audit.api.BuiltinLogEntryData.LOG_EVENT_ID;
+import static org.nuxeo.ecm.platform.audit.api.BuiltinLogEntryData.LOG_ID;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.junit.Test;
+import org.nuxeo.ecm.core.api.ScrollResult;
+import org.nuxeo.ecm.core.query.sql.model.Literals;
+import org.nuxeo.ecm.core.query.sql.model.Operator;
+import org.nuxeo.ecm.core.query.sql.model.Predicate;
+import org.nuxeo.ecm.core.query.sql.model.Reference;
+import org.nuxeo.ecm.platform.audit.api.AuditQueryBuilder;
+import org.nuxeo.ecm.platform.audit.api.LogEntry;
+import org.nuxeo.ecm.platform.audit.api.Predicates;
+import org.nuxeo.ecm.platform.audit.impl.LogEntryImpl;
+import org.nuxeo.ecm.platform.audit.service.AbstractAuditBackend;
+import org.nuxeo.ecm.platform.audit.service.AuditBackend;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+/**
+ * Audit test that are common across all storage backends
+ */
+public abstract class AbstractAuditStorageTest {
+
+    public static final String ID_FOR_AUDIT_STORAGE_TESTS = "ID_FOR_AUDIT_STORAGE_TESTS";
+
+    public static final int NUM_OF_EVENTS = 56;
+
+    @Inject
+    protected AuditBackend auditBackend;
+
+    protected static ObjectMapper mapper = new ObjectMapper();
+
+    /**
+     * Sets up data for the tests below. Doesn't run using @Before so as not to interfere with other audit backends
+     */
+    public void setUpTestData() throws Exception {
+
+        AbstractAuditBackend backend = (AbstractAuditBackend) this.auditBackend;
+        AuditQueryBuilder builder = new AuditQueryBuilder().predicates(
+                Predicates.eq(LOG_EVENT_ID, ID_FOR_AUDIT_STORAGE_TESTS));
+        List<LogEntry> logs = backend.queryLogs(builder);
+
+        if (logs.isEmpty()) {
+
+            List<String> jsonEntries = new ArrayList<>();
+
+            for (int i = 1; i <= NUM_OF_EVENTS; i++) {
+                ObjectNode logEntryJson = mapper.createObjectNode();
+                logEntryJson.put(LOG_ID, 1000L+Integer.valueOf(i).longValue());
+                logEntryJson.put(LOG_EVENT_ID, ID_FOR_AUDIT_STORAGE_TESTS);
+                logEntryJson.put(LOG_DOC_PATH, i % 2 == 0 ? "/is/even" : "/is/odd");
+                jsonEntries.add(mapper.writeValueAsString(logEntryJson));
+            }
+
+            backend.append(jsonEntries);
+            flush();
+        }
+    }
+
+    protected abstract void flush() throws Exception;
```

Comment:
I think you should rely on `CoreFeature.waitForAsyncCompletion` which can handle waiters. You have an example in `AuditFeature.BulkAuditWaiter`. This approach is really nice, cause each layers can add its waiters, you can do specific call to wait the whole stack and it will also be taken into account when framework needs to deploy/undeploy stuffs between tests.
But for now, keep the `flush` method, I'm afraid work to have a generic audit waiter is a bit huge.

Reply:
Even if `CoreFeature#waitForAsyncCompletion` calls `TransactionalFeature#nextTransaction`, I would go for explicitly using the `TransactionalFeature` if the only thing we want is starting a new transaction (and we don't care about the async completion):
```
@Inject
protected TransactionalFeature txFeature;
...
txFeature.nextTransaction();
```